### PR TITLE
Add free/total inode counts

### PR DESCRIPTION
**Pull request by Space2Man**
Hi,

initial pull request ... not sure about the variable names ... in LibC I have called them (free|total)InodeCounts because that's what it  is (at least on Linux) ... not sure if it's called like that for all the other OS as well. In FreeBSD it's called

> file serial numbers (i.e., inodes)

But in LinuxFileSystem I have called them file counts ... let me know what you prefer. Furthermore I had issues using the "new" JNA stuff ... if you convert the existing Sysinfo I can adapt Statvfs as well.

Furthermore I have set it to -1 for all other OS except of Linux ... not sure how to implement it for FreeBSD, Solaris, MacOS, ...

### CODE REVIEW

File: oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java [hunk 1]
```diff
@@ -188,7 +188,28 @@ private boolean listElementStartsWith(List<String> aList, String charSeq) {
...
+
+            try {
+                Libc.Statvfs vfsStat = new Libc.Statvfs();
+                if (0 == Libc.INSTANCE.statvfs(path, vfsStat)) {
+                    totalFiles = vfsStat.fsTotalInodeCount.longValue();
+                    usableFiles = vfsStat.fsFreeInodeCount.longValue();
+                    totalSpace = vfsStat.fsSizeInBlocks.longValue() * vfsStat.fsBlockSize.longValue();
+                    usableSpace = vfsStat.fsBlocksFree.longValue() * vfsStat.fsBlockSize.longValue();
+                } else {
+                    totalSpace = new File(path).getTotalSpace();
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java]: don't create a new `File` object twice.

File: oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java [hunk 2]
```diff
@@ -188,7 +188,28 @@ private boolean listElementStartsWith(List<String> aList, String charSeq) {
...
+                    usableSpace = new File(path).getUsableSpace();
+                    LOG.error("Failed to get statvfs. Error code: {}", Native.getLastError());
+                }
+            } catch (UnsatisfiedLinkError | NoClassDefFoundError e) {
+                LOG.error("Failed to get file counts from statvfs. {}", e);
+            }
+
+            OSFileStore osStore = new OSFileStore(name, volume, path, description, type, uuid, usableSpace, totalSpace, usableFiles, totalFiles);
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java]: Use a no-argument constructor and setters.

File: oshi-json/src/main/java/oshi/json/software/os/OSFileStore.java [hunk 3]
```diff
@@ -241,6 +245,44 @@ public void setTotalSpace(long value) {
         this.fileStore.setTotalSpace(value);
     }
 
+    /**
+     * Usable files / free inodes on the drive.
+     *
+     * @return Usable files / free inodes on the drive (count)
+     */
+    public long getUsableFiles() {
```
Comment (dbwiddis) [inline oshi-json/src/main/java/oshi/json/software/os/OSFileStore.java]: per comments, let's go with freeInodes()

**Comment by dbwiddis** [comment]
Looking good so far....

For Solaris you should be able to do [something similar to Linux with the Statvfs](https://docs.oracle.com/cd/E26502_01/html/E29032/statvfs-2.html) structure.   For Mac and FreeBSD you're going to have to parse the output of `df -i`.  And for Windows just leave as default/unimplemented.

**Comment by dbwiddis** [comment]
Also travis-ci can no longer test on `oraclejdk10` so for the time being you can just delete that in your `.travis.yml` file so the PR build tests don't show red.  I can't seem to get it to work with `oraclejdk11` yet due to some issue with surefire plugin and writing to the console.  Hopefully I'll get it figured out soon...

**Comment by Space2Man** [comment]
Is someone able to help with Solaris / Mac / FreeBSD? I have no such system at hand ...

**Comment by dbwiddis** [comment]
I can help with those.  Can you get rid of the “-1, -1” add ones to those constructors though? Stick with the original constructors or switch to no-arg with the -1 as the initialization.

**Comment by dbwiddis** [comment]
Looks good. I'll see if I can check out your branch and send commits to it for the other OS's.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I'm committing to your master branch, which is sometimes showing up here but then disappearing from the pull request.  When I'm done you may need to pull and push to get this PR to include my changes.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I'm committing to your master branch, which is sometimes showing up here but then disappearing from the pull request.  When I'm done you may need to pull and push to get this PR to include my changes.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I'm committing to your master branch, which is sometimes showing up here but then disappearing from the pull request.  When I'm done you may need to pull and push to get this PR to include my changes.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I'm committing to your master branch, which is sometimes showing up here but then disappearing from the pull request.  When I'm done you may need to pull and push to get this PR to include my changes.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I'm committing to your master branch, which is sometimes showing up here but then disappearing from the pull request.  When I'm done you may need to pull and push to get this PR to include my changes.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github is having some issues tonight](https://status.github.com/messages) so I'll check again in the morning.

**Comment by dbwiddis** [comment]
Heh. Mac was trivial, the needed values were already in the `Statfs` structure we already had.

>  Macintosh HD (Volume) [apfs] 86.4 GiB of 465.7 GiB free (18.6%), 9223372036852193649 of 9223372036854775807 files free (100.0%) is /dev/disk1s1 [/] and is mounted at null

I've committed implementations for Mac, Solaris, and FreeBSD to your master branch.  They're not showing up here yet but [github was having some issues tonight](https://status.github.com/messages).

File: oshi-core/src/main/java/oshi/jna/platform/linux/Libc.java [hunk 4]
```diff
@@ -51,15 +58,107 @@
...
+        public NativeLong fsBlockSize;
+        public NativeLong fsFragmentSize;
+        public NativeLong fsSizeInBlocks;
+        public NativeLong fsBlocksFree;
+        public NativeLong fsBlocksFreeUnpriv;
+        public NativeLong fsTotalInodeCount;
+        public NativeLong fsFreeInodeCount;
+        public NativeLong fsFreeInodeCountUnpriv;
+        public NativeLong fsId;
+        public int[] _fPad32bit = new int[PADDING_SIZE]; // 0 on 64-bit systems
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/jna/platform/linux/Libc.java]: So I'm looking through docs everywhere and can't find why this padding exists. Can you point me to the source you used to determine this structure?

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/jna/platform/linux/Libc.java]: Ah, nevermind. [Found it here. ](https://code.woboq.org/gcc/include/bits/statvfs.h.html) We should name it `_f_unused` though.

**Comment by dbwiddis** [comment]
Leaving a note for myself on what I need to do tonight:
- Fix whatever's breaking macOS tests.  There's a null somewhere, probably related to not setting the logical drive (to fix, should probably initialize to empty string).
- Check if the same error on macOS is on Solaris and FreeBSD (it probably is but initializing to empty string would fix)
- Add the inode values to the unit tests (with special case for Windows)
- write a test for new structure/function in Linux LibC (and possibly submit that new structure/function to JNA)

**Comment by dbwiddis** [comment]
No... that error was everywhere and I backed it out the other 4 OS's but since we weren't adding the new fields to Windows I didn't even bother looking! Whoops. Can you fix?

**Pull request merged**